Add X.509 thumbprint client identifier scheme

martijnharing opened the pull request:
Implements  #320 : add x509_x5t, copied text from other client identifier schemes where applicable.

bc-pi reviewed (changes requested):
> If the Wallet can establish trust in the Client Identifier authenticated through the certificate, e.g. because the leaf X.509 certificate is signed by a trusted X.509 certificate.
is missing the latter part of the typical 'If [condition], then [result]' construction. 

x509_x5t is probably not a good name https://github.com/openid/OpenID4VP/issues/320#issuecomment-2663163027 

and I had thought we had "agreed to drop x509_san_uri" https://github.com/openid/OpenID4VP/issues/320#issuecomment-2652058132

martijnharing commented:
> > If the Wallet can establish trust in the Client Identifier authenticated through the certificate, e.g. because the leaf X.509 certificate is signed by a trusted X.509 certificate.
> > is missing the latter part of the typical 'If [condition], then [result]' construction.
> 
> x509_x5t is probably not a good name [#320 (comment)](https://github.com/openid/OpenID4VP/issues/320#issuecomment-2663163027)
> 
> and I had thought we had "agreed to drop x509_san_uri" [#320 (comment)](https://github.com/openid/OpenID4VP/issues/320#issuecomment-2652058132)

Changed to x509_hash and removed x509_san_uri and updated relevant examples to x509_san_dns (hopefully correctly)

c2bo commented on openid-4-verifiable-presentations-1_0.md:
```diff
@@ -572,11 +572,11 @@ Body
 * `web-origin`: This Client Identifier Scheme is defined in (#dc_api_request). The Wallet MUST NOT accept this Client Identifier Scheme if the request is not sent via the Digital Credentials API defined in (#dc_api).
 
-To use the Client Identifier Schemes `https`, `did`, `verifier_attestation`, `x509_san_dns`, and `x509_san_uri`, Verifiers MUST be confidential clients. This might require changes to the technical design of native apps as such apps are typically public clients.
+* `x509_hash`: When the Client Identifier Scheme is `x509_hash`, the Client Identifier MUST be a thumbprint and match a tuhmbprint in leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header [@!RFC7515] of the signed request object.  ...
```
```suggestion
* `x509_hash`: When the Client Identifier Scheme is `x509_hash`, the Client Identifier MUST be a thumbprint and match the thumbprint in the leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header parameter [@!RFC7515] of the signed request object. The value of `x509_hash` is the base64url encoded value of the SHA-256 hash of the DER-encoded X.509 certificate. The Wallet MUST validate the signature and the trust chain of the X.509 certificate. All verifier metadata other than the public key MUST be obtained from the `client_metadata` parameter. If the Wallet can establish trust in the Client Identifier authenticated through the certificate, e.g. because the leaf X.509 certificate is signed by a trusted X.509 certificate. Example Client Identifier: `x509_hash:Uvo3HtuIxuhC92rShpgqcT3YXwrqRxWEviRiA0OZszk`
```
This needs some adjustment I do believe? I do believe we can remove it?
> If the Wallet can establish trust in the Client Identifier authenticated through the certificate, e.g. because the leaf X.509 certificate is signed by a trusted X.509 certificate.

bc-pi replied:
as previously mentioned in https://github.com/openid/OpenID4VP/pull/430#pullrequestreview-2621192806 the sentence, "If the Wallet can establish trust in the Client Identifier authenticated through the certificate, e.g. because the leaf X.509 certificate is signed by a trusted X.509 certificate." is missing the latter part of the typical 'If [condition], then [result]' construction that one would expect in a sentence. If there's a then there, then the then should be added to the text. If there's not a then there, then the whole sentence should be removed.

bc-pi commented:
> updated relevant examples to x509_san_dns (hopefully correctly) 

did you search for all occurrences of `x509_san_dns` in the repo?

c2bo commented on openid-4-verifiable-presentations-1_0.md:
```diff
@@ -572,11 +572,11 @@ Body
 * `web-origin`: This Client Identifier Scheme is defined in (#dc_api_request). The Wallet MUST NOT accept this Client Identifier Scheme if the request is not sent via the Digital Credentials API defined in (#dc_api).
 
-To use the Client Identifier Schemes `https`, `did`, `verifier_attestation`, `x509_san_dns`, and `x509_san_uri`, Verifiers MUST be confidential clients. This might require changes to the technical design of native apps as such apps are typically public clients.
+* `x509_hash`: When the Client Identifier Scheme is `x509_hash`, the Client Identifier MUST be a thumbprint and match the thumbprint in the leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header parameter [@!RFC7515] of the signed  ...
```
```suggestion
* `x509_hash`: When the Client Identifier Scheme is `x509_hash`, the Client Identifier MUST be a thumbprint and match the thumbprint of the leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header parameter [@!RFC7515] of the signed request object. The value of `x509_hash` is the base64url encoded value of the SHA-256 hash of the DER-encoded X.509 certificate. The Wallet MUST validate the signature and the trust chain of the X.509 certificate. All verifier metadata other than the public key MUST be obtained from the `client_metadata` parameter. Example Client Identifier: `x509_hash:Uvo3HtuIxuhC92rShpgqcT3YXwrqRxWEviRiA0OZszk`
```

bc-pi replied:
agree w/ @c2bo

c2bo commented:
examples/response/jwt_vp.json contains x509_san_uri in the example (and is used in the spec) -> would need to change that

bc-pi commented:
> examples/response/jwt_vp.json contains x509_san_uri in the example (and is used in the spec) -> would need to change that

yes, that's one one would find when search for all occurrences of x509_san_dns in the repo

paulbastian commented on openid-4-verifiable-presentations-1_0.md:
```diff
@@ -572,11 +572,11 @@ Body
 * `web-origin`: This Client Identifier Scheme is defined in (#dc_api_request). The Wallet MUST NOT accept this Client Identifier Scheme if the request is not sent via the Digital Credentials API defined in (#dc_api).
 
-To use the Client Identifier Schemes `https`, `did`, `verifier_attestation`, `x509_san_dns`, and `x509_san_uri`, Verifiers MUST be confidential clients. This might require changes to the technical design of native apps as such apps are typically public clients.
+* `x509_hash`: When the Client Identifier Scheme is `x509_hash`, the Client Identifier MUST be a thumbprint and match the thumbprint in the leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header parameter [@!RFC7515] of the signed  ...
```
This is different from X509 fingerprint/thumbprint, that is done with SHA-1. This may lead to confusion? Maybe avoid the words "thumbprint"

paulbastian commented on openid-4-verifiable-presentations-1_0.md:
```diff
@@ -572,11 +572,11 @@ Body
 * `web-origin`: This Client Identifier Scheme is defined in (#dc_api_request). The Wallet MUST NOT accept this Client Identifier Scheme if the request is not sent via the Digital Credentials API defined in (#dc_api).
 
-To use the Client Identifier Schemes `https`, `did`, `verifier_attestation`, `x509_san_dns`, and `x509_san_uri`, Verifiers MUST be confidential clients. This might require changes to the technical design of native apps as such apps are typically public clients.
+* `x509_hash`: When the Client Identifier Scheme is `x509_hash`, the Client Identifier MUST be a hash and match the hash of the leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header parameter [@!RFC7515] of the signed request obje ...
```
```suggestion
* `x509_hash`: When the Client Identifier Scheme is `x509_hash`, the Client Identifier MUST be a hash and match the hash of the leaf certificate passed with the request. The request MUST be signed with the private key corresponding to the public key in the leaf X.509 certificate of the certificate chain added to the request in the `x5c` JOSE header parameter [@!RFC7515] of the signed request object. The value of `x509_hash` is the base64url encoded value of the SHA-256 hash of the DER-encoded X.509 certificate. The Wallet MUST validate the signature and the trust chain of the X.509 leaf certificate. All verifier metadata other than the public key MUST be obtained from the `client_metadata` parameter. Example Client Identifier: `x509_hash:Uvo3HtuIxuhC92rShpgqcT3YXwrqRxWEviRiA0OZszk`
```
